lib/cli/src/Cardano/CLI.hs

```diff
@@ -395,189 +337,428 @@ runCliCommand Environment {..} manager execServe execLaunch
-    -> Maybe (Hash "Genesis")
-    -> Command
-commandWalletServe
-        name listen backendPort stateDir network verbosity mGenesisHash =
-    Command name args (threadDelay oneSecond) Inherit
+{-------------------------------------------------------------------------------
+                              Options & Arguments
+-------------------------------------------------------------------------------}
+
+-- | --state=STRING
+addressStateOption :: Parser AddressState
+addressStateOption = optionT $ mempty
+    <> long "address"
+    <> metavar "STRING"
+    <> help "only addresses with the given state: either 'used' or 'unused'."
+
+-- | --database=FILEPATH
+databaseOption :: Parser FilePath
+databaseOption = optionT $ mempty
+    <> long "database"
+    <> metavar "FILEPATH"
+    <> help "use this file for storing wallet state. Run in-memory otherwise."
+
+-- | [--random-port|--port=INT]
+listenOption :: Parser Listen
+listenOption =
+    (ListenOnRandomPort <$ randomPortOption)
+    <|>
+    (ListenOnPort . getPort <$> portOption)
+
+-- | [--random-port]
+randomPortOption :: Parser Bool
+randomPortOption = flag' False $ mempty
+    <> long "random-port"
+    <> help "serve wallet API on any available port (conflicts with --port)"
+
+-- | [--node-port=INT], default: 8080
+nodePortOption :: Parser (Port "Node")
+nodePortOption = optionT $ mempty
+    <> long "node-port"
+    <> metavar "INT"
+    <> help "port used for communicating with the target node."
+    <> value (Port 8080)
+    <> showDefaultWith showT
+
+-- | --payment=PAYMENT
+paymentOption :: DecodeAddress t => Parser (AddressAmount t)
+paymentOption = optionT $ mempty
+    <> long "payment"
+    <> metavar "PAYMENT"
+    <> help
+        "address to send to and amount to send separated by @\
+        \, e.g. '<amount>@<address>'"
+
+-- | [--address-pool-gap=INT], default: 20
+poolGapOption :: Parser AddressPoolGap
+poolGapOption = optionT $ mempty
+    <> long "address-pool-gap"
+    <> metavar "INT"
+    <> help "number of unused consecutive addresses to keep track of."
+    <> value defaultAddressPoolGap
+    <> showDefaultWith showT
+
+-- | [--port=INT], default: 8090
+portOption :: Parser (Port "Wallet")
+portOption = optionT $ mempty
+    <> long "port"
+    <> metavar "INT"
+    <> help "port used for serving the wallet API."
+    <> value (Port 8090)
+    <> showDefaultWith showT
+
+-- | [--size=INT], default: 15
+sizeOption :: Parser Int
+sizeOption = optionT $ mempty
+    <> long "size"
+    <> metavar "INT"
+    <> help "number of mnemonic words to generate."
+    <> value 15
+    <> showDefaultWith show
+
+-- | --state-dir=FILEPATH
+stateDirOption :: Parser FilePath
+stateDirOption = optionT $ mempty
+    <> long "state-dir"
+    <> metavar "DIR"
+    <> help "write wallet state (blockchain and database) to this directory"
+
+-- | [(--quiet|--verbose)]
+verbosityOption :: Parser Verbosity
+verbosityOption = (Quiet <$ quiet) <|> (Verbose <$ verbose) <|> (pure Default)
```

Comment:
If I type:
```hs
cardano-wallet serve --verbose --quiet
```
It gives me the following error:
```
Invalid option `--quiet'
```
I wonder if `optparse-applicative` provides a way to produce a better error here. For example:
```
The options `--verbose` and `--quiet` are mutually exclusive.
```
(or something similar.)

I know this is a fairly trivial example (and therefore probably doesn't matter very much), but it would be nice to know if there's a solution.

Reply:
I can have a look indeed :thinking: ... The tough part with `optparse-applicative` is that all the parsers are _just_ `Applicative` (and `Alternative`) but we don't have any `Monad` instances which makes chaining them a bit complex. Therefore, control-flows are quite tricky ^^"

Reply:
I've just seen that someone posed a similar question here: https://stackoverflow.com/questions/30224307/how-do-you-provide-a-specific-error-message-with-optparse-applicative-when-multiple-mutually-exclusive-options-are-provided

Perhaps we could use a similar solution?

Reply:
Well, the major downside of this solution is that we end up with an artificial type that we have to carry up to each command handler and mix the parsing logic with the execution of the command itself.. :/

Reply:
Yes, after looking in more detail, there really doesn't seem to be a clean solution to this.

(Of course, one could change the return type to `Parser (Either VerbosityError Verbosity)`, but then this function becomes awkward to use in a context which expects `Parser Verbosity`, and since there isn't a `Monad` instance...)